### PR TITLE
Use KfError interface

### LINKED ISSUE #2841: Convert kfctl error to kferrors interface.
In order to integrate deploy app and kfctl code, and have monitoring signals properly send back, we need to make following functions return kferror interface:

gcp.go:
* generate()
* apply()

ksonnet.go:
* generate()
* apply()

kferrors interface:
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/pkg/apis/kferrors.go

### PR DESCRIPTION
**Pull request by gabrielwen**
Fixes #2841

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2854)

### CODE REVIEW

**Comment by kunmingg** [comment]
Seems ```INTERNAL_ERROR``` is being used too widely.
```INTERNAL_ERROR``` from monitoring likely means we need to patch fix and release asap.

Can we label errors like:
* INTERNAL_ERROR: there's likely a bug in kfctl and we need to make code change.
* INVALID_ARGUMENT: error is caused by user's input or returned from GCP APIs. kfctl don't need to change.
* UNKNOWN: used only when the error doesn't belong to above 2 cases. should be rare.

**Comment by kunmingg** [comment]
And can we make following functions return ```KfError```?
So we always have a ```StatusCode``` when error returned? 

gcp.go:
generate()
apply()
ksonnet.go:
generate()
apply()

**Comment by gabrielwen** [comment]
> And can we make following functions return `KfError`?
> So we always have a `StatusCode` when error returned?
> 
> gcp.go:
> generate()
> apply()
> ksonnet.go:
> generate()
> apply()

I've made changes to all files under bootstrap/pkg/...

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 1]
```diff
@@ -569,24 +692,34 @@ func (gcp *Gcp) Apply(resources kftypes.ResourceEnum) error {
 	if gcp.isCLI {
 		if gcp.Spec.UseBasicAuth {
 			if os.Getenv(kftypes.KUBEFLOW_USERNAME) == "" || os.Getenv(kftypes.KUBEFLOW_PASSWORD) == "" {
-				return fmt.Errorf("gcp apply needs ENV %v and %v set when using basic auth",
-					kftypes.KUBEFLOW_USERNAME, kftypes.KUBEFLOW_PASSWORD)
+				return &kfapis.KfError{
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Can we have generate() and apply() return kfapis.KfError?
Like ```func (gcp *Gcp) Apply(resources kftypes.ResourceEnum) kfapis.KfError```

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 2]
```diff
@@ -1025,11 +1217,17 @@ func (gcp *Gcp) createGcpServiceAcctSecret(ctx context.Context, client *clientse
 	}
 	saKey, err := iamService.Projects.ServiceAccounts.Keys.Create(name, req).Context(ctx).Do()
 	if err != nil {
-		return fmt.Errorf("Service account key creation error: %v", err)
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Switch to "INVALID_ARGUMENT"?
We are purely depend on iam API here so might not be responsible for error here.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 3]
```diff
@@ -1189,7 +1439,10 @@ func (gcp *Gcp) gcpInitProject() error {
 		service := fmt.Sprintf("projects/%v/services/%v", gcp.Spec.Project, api)
 		_, opErr := serviceusageService.Services.Enable(service, &serviceusage.EnableServiceRequest{}).Context(ctx).Do()
 		if opErr != nil {
-			return fmt.Errorf("could not enable API service %v: %v", api, opErr)
+			return &kfapis.KfError{
+				Code:    int(kfapis.INTERNAL_ERROR),
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Change to "INVALID_ARGUMENT" here?

File: bootstrap/pkg/kfapp/ksonnet/ksonnet.go [hunk 4]
```diff
@@ -162,14 +177,21 @@ func (ksApp *ksApp) applyComponent(components []string, cfg *clientcmdapi.Config
...
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("%v failed components in last try", len(components)-len(doneApply)),
+		}
 	}, bo)
 	if err != nil {
 		log.Errorf("components apply failed; Error: %v", err)
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: err.Error(),
```
Comment (kunmingg) [inline bootstrap/pkg/kfapp/ksonnet/ksonnet.go]: Can we keep error message:
("components apply failed; Error: %v"...

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 5]
```diff
@@ -481,62 +547,110 @@ func (gcp *Gcp) updateDM(resources kftypes.ResourceEnum) error {
 	ctx := context.Background()
 	gcpClient := oauth2.NewClient(ctx, gcp.tokenSource)
 	if err := gcp.updateDeployment(gcp.Name+"-storage", STORAGE_FILE); err != nil {
-		return fmt.Errorf("could not update %v: %v", STORAGE_FILE, err)
+		return &kfapis.KfError{
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: sure.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 6]
```diff
@@ -596,12 +729,18 @@ func (gcp *Gcp) Apply(resources kftypes.ResourceEnum) error {
 	// Update deployment manager
 	updateDMErr := gcp.updateDM(resources)
 	if updateDMErr != nil {
-		return fmt.Errorf("gcp apply could not update deployment manager Error %v", updateDMErr)
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: returning code from `updateDM`, let's use follow up PR for `KfError`.  it'll change a lot of things and we probably want to use a different PR for it.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 7]
```diff
@@ -596,12 +729,18 @@ func (gcp *Gcp) Apply(resources kftypes.ResourceEnum) error {
...
+			Message: fmt.Sprintf("gcp apply could not update deployment manager Error %v", updateDMErr),
+		}
 	}
 	// Insert secrets into the cluster
 	secretsErr := gcp.createSecrets()
 	if secretsErr != nil {
-		return fmt.Errorf("gcp apply could not create secrets Error %v", secretsErr)
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 8]
```diff
@@ -895,23 +1070,23 @@ func (gcp *Gcp) writeStorageConfig(src string, dest string) error {
 	buf, err := ioutil.ReadFile(src)
 	if err != nil {
 		return &kfapis.KfError{
-			Code:    int(kfapis.INTERNAL_ERROR),
+			Code:    int(kfapis.INVALID_ARGUMENT),
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 9]
```diff
@@ -895,23 +1070,23 @@ func (gcp *Gcp) writeStorageConfig(src string, dest string) error {
...
 			Message: fmt.Sprintf("Error when reading storage-kubeflow template: %v", err),
 		}
 	}
 
 	var data map[string]interface{}
 	if err = yaml.Unmarshal(buf, &data); err != nil {
 		return &kfapis.KfError{
-			Code:    int(kfapis.INTERNAL_ERROR),
+			Code:    int(kfapis.INVALID_ARGUMENT),
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done.

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 10]
```diff
@@ -895,23 +1070,23 @@ func (gcp *Gcp) writeStorageConfig(src string, dest string) error {
...
 			Message: fmt.Sprintf("Error when unmarshaling template %v: %v", src, err),
 		}
 	}
 
 	res, ok := data["resources"]
 	if !ok {
 		return &kfapis.KfError{
-			Code:    int(kfapis.INTERNAL_ERROR),
+			Code:    int(kfapis.INVALID_ARGUMENT),
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: could be - it's from `storage-kubeflow.yaml`, so it's user input if user changes it.

File: bootstrap/pkg/kfapp/gcp/gcp.go (hunk 2 above)
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done.

File: bootstrap/pkg/kfapp/gcp/gcp.go (hunk 3 above)
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done.

File: bootstrap/pkg/kfapp/ksonnet/ksonnet.go (hunk 4 above)
Comment (gabrielwen) [reply bootstrap/pkg/kfapp/ksonnet/ksonnet.go]: done.

**Pull request merged**